semantic/poly_types.go

```diff
@@ -132,23 +148,23 @@ func (tv Tvar) Equal(t PolyType) bool {
 
 // PolyType methods for Nature
 func (n Nature) occurs(Tvar) bool                                { return false }
-func (n Nature) substituteType(Tvar, PolyType) PolyType          { return n }
+func (n Nature) apply(sub Substitution) (PolyType, bool)         { return n, false }
 func (n Nature) resolveType(map[Tvar]Kind) (Type, error)         { return n, nil }
 func (n Nature) MonoType() (Type, bool)                          { return n, true }
 func (n Nature) resolvePolyType(map[Tvar]Kind) (PolyType, error) { return n, nil }
 func (n Nature) freeVars(*Constraints) TvarSet                   { return nil }
-func (n Nature) unifyType(kinds map[Tvar]Kind, t PolyType) (Substitution, error) {
+func (n Nature) unifyType(kinds map[Tvar]Kind, s Substitution, t PolyType) (Substitution, error) {
 	switch t := t.(type) {
 	case Nature:
 		if t != n {
 			return nil, fmt.Errorf("%v != %v", n, t)
 		}
 	case Tvar:
-		return t.unifyType(kinds, n)
+		return t.unifyType(kinds, s, n)
 	default:
 		return nil, fmt.Errorf("cannot unify %v with %T", n, t)
 	}
-	return nil, nil
+	return s, nil
```

Comment:
Why return `s` here?  If you get here, you have two equal `Nature` types, so the substitution should be `nil`?

Reply:
The `unifyTypes` method now takes the substitution for the whole program as input, and returns a possibly updated program substitution. In this case if the two types unify, we just return the substitution that was passed in.

Only when we unify a type variable with another type is there a possibility that we return an expanded substitution.

Reply:
I see... the comment on the interface method should be updated to reflect that.